app/app/src/main/java/com/example/myapplication/MainActivity.java

```diff
@@ -410,6 +423,52 @@ void goBackward() throws IOException {
         int msg = 5;
         Car.mmOutputStream.write(msg);
     }
+
+    // joyStick control methods to replace four buttons
+    public void onMove(JoyStick joyStick, double angle, double power, int direction) {
+        switch (direction){
+            case JoyStick.DIRECTION_LEFT:
+                Toast.makeText(MainActivity.this,"Go Left",Toast.LENGTH_SHORT).show();
+                try {
+                    goLeft();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+                break;
+            case JoyStick.DIRECTION_RIGHT:
+                Toast.makeText(MainActivity.this,"Go Right",Toast.LENGTH_SHORT).show();
+                try {
+                    goRight();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+                break;
+            case JoyStick.DIRECTION_UP:
+                Toast.makeText(MainActivity.this,"Go Forward",Toast.LENGTH_SHORT).show();
+                try {
+                    goForward();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+                break;
+            case JoyStick.DIRECTION_DOWN:
+                Toast.makeText(MainActivity.this,"Go back",Toast.LENGTH_SHORT).show();
+                try {
+                    goBackward();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+                break;
+            default:
+                break;
+        }
+    }
+    public void onTap() {
+
+    }
+    public void onDoubleTap() {
```

Comment:
What are these empty methods for?

Reply:
It generates automatically when I implement with JoyStick.JoyStickListener